compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUConfigureTensorLayouts.cpp

```diff
@@ -79,102 +80,138 @@ static int64_t getSubgroupNCount(Operation *op) {
-///
-/// And here inner most is referential to the iteration order, not the order
-/// they appear per fragment (because there is no relationship between the
-/// dimension order of M in A and in C, for example).
-static NestedLayoutAttr createNestedLayout(
-    MLIRContext *context, int64_t rank, int64_t outerDim, int64_t innerDim,
-    ArrayRef<int64_t> subgroupSizes, ArrayRef<int64_t> subgroupStrides,
-    ArrayRef<int64_t> batchCount, IREE::GPU::MMASingleSubgroupLayout counts) {
-
-  LLVM_DEBUG({
-    llvm::dbgs() << "Creating Nested Layout for::" << "\n    outerDim = "
-                 << outerDim << "\n    innerDim = " << innerDim
-                 << "\n    subgroupSizes: " << llvm::interleaved(subgroupSizes)
-                 << "\n    subgroupStrides: "
-                 << llvm::interleaved(subgroupStrides)
-                 << "\n    batchCount: " << llvm::interleaved(batchCount)
-                 << "\n    counts.outer: " << llvm::interleaved(counts.outer)
-                 << "\n    counts.thread: " << llvm::interleaved(counts.thread)
-                 << "\n    counts.element: "
-                 << llvm::interleaved(counts.element)
-                 << "\n    counts.tstrides: "
-                 << llvm::interleaved(counts.tstrides) << "\n";
-  });
+  int64_t rank = bounds.size();
+  FailureOr<VectorContractOpInfo> maybeOpInfo =
+      VectorContractOpInfo::inferFromIndexingMaps(contractIndexingMaps);
+  if (failed(maybeOpInfo)) {
+    return failure();
+  }
+  VectorContractOpInfo opInfo = maybeOpInfo.value();
+  // Get the inner dimensions.
+  int64_t innerMDim = opInfo.getMDims().back();
+  int64_t innerNDim = opInfo.getNDims().back();
+  int64_t innerKDim = opInfo.getKDims().back();
+  // Find the number Of subgroups being used from subgroupMCount/subgroupNCount.
+  // Just assign them to the inner-most dimension for now.
+  // TODO: Use subgroup_basis to get this instead and allow distributing
+  // subgroups on multiple dimensions.
+  SmallVector<int64_t> subgroupCounts(rank, 1);
+  SmallVector<int64_t> subgroupStrides(rank, 0);
+  subgroupCounts[innerMDim] = subgroupMCount;
+  subgroupCounts[innerNDim] = subgroupNCount;
+  // Distribute on M and then N.
+  // TODO: Use subgroup_basis to get the strides.
+  subgroupStrides[innerMDim] = 1;
+  subgroupStrides[innerNDim] = subgroupNCount;
+
+  // Since these MMA intrinsics have a given tile size for each subgroup, we can
+  // calculate the batch dimensions without looking at the subgroup layout.
+  SmallVector<int64_t> subgroupSize(rank, 1);
+  auto [mSize, nSize, kSize] = mmaIntrinsic.getMNKShape();
+  subgroupSize[innerMDim] = mSize;
+  subgroupSize[innerNDim] = nSize;
+  subgroupSize[innerKDim] = kSize;
+
+  SmallVector<int64_t> batchCounts(rank);
+  for (auto i : llvm::seq<int64_t>(rank)) {
+    int64_t workgroupDimSize = subgroupCounts[i] * subgroupSize[i];
+    batchCounts[i] = llvm::divideCeil(bounds[i], workgroupDimSize);
+  }
```

Comment:
optional: You could also do
```suggestion
  SmallVector<int64_t> batchCounts(rank);
  for (auto [batchCount, subgroupCount, subgroupSize, bound] : llvm::zip_equal(batchCounts, subgroupCounts, subgroupSize, bounds)) {
    int64_t workgroupDimSize = subgroupCount * subgroupSize;
    batchCount = llvm::divideCeil(bound, workgroupDimSize);
  }
```